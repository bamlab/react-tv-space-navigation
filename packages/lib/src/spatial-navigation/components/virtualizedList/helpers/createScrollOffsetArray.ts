import { ScrollBehavior } from '../VirtualizedList';
import { computeTranslation } from './computeTranslation';
import { getLastLeftItemIndex, getLastRightItemIndex } from './getLastItemIndex';

/**
 * This function precomputes all scroll offsets
 * It won't move until data moves or the itemSize changes
 */
export const computeAllScrollOffsets = <T>({
  itemSize,
  nbMaxOfItems,
  numberOfItemsVisibleOnScreen,
  scrollBehavior,
  data,
  listSizeInPx,
}: {
  itemSize: number | ((item: T) => number);
  nbMaxOfItems: number;
  numberOfItemsVisibleOnScreen: number;
  scrollBehavior: ScrollBehavior;
  data: T[];
  listSizeInPx: number;
}) => {
  const maxPossibleLeftAlignedIndex = getLastLeftItemIndex<T>(data, itemSize, listSizeInPx);
  const maxPossibleRightAlignedIndex = getLastRightItemIndex<T>(data, itemSize, listSizeInPx);

  const scrollOffsets = data.map((_, index) =>
    computeTranslation({
      currentlyFocusedItemIndex: index,
      itemSizeInPx: itemSize,
      nbMaxOfItems: nbMaxOfItems ?? data.length,
      numberOfItemsVisibleOnScreen: numberOfItemsVisibleOnScreen,
      scrollBehavior: scrollBehavior,
      data: data,
      listSizeInPx: listSizeInPx,
      maxPossibleLeftAlignedIndex: maxPossibleLeftAlignedIndex,
      maxPossibleRightAlignedIndex: maxPossibleRightAlignedIndex,
    }),
  );

  return scrollOffsets;
};
