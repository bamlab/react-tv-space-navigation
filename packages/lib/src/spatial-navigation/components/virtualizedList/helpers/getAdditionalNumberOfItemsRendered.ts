import { ScrollBehavior } from '../VirtualizedList';

export const getAdditionalNumberOfItemsRendered = (
  scrollBehavior: ScrollBehavior,
  numberOfElementsVisibleOnScreen: number,
) => {
  switch (scrollBehavior) {
    case 'stick-to-start':
    case 'stick-to-end':
      return 4 + numberOfElementsVisibleOnScreen;
    case 'jump-on-scroll':
      return 2 * numberOfElementsVisibleOnScreen + 1;
  }
};
