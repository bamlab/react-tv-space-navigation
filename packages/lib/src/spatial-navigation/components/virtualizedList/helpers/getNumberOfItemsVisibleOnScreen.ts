const getMinSizeOfItems = <T>({
  data,
  itemSize,
}: {
  data: T[];
  itemSize: number | ((item: T) => number);
}) => {
  if (typeof itemSize === 'number') {
    return itemSize;
  }

  if (data.length === 0) {
    return 0;
  }

  const firstElementSize = itemSize(data[0]);

  const minSize = data.reduce<number>((smallestSize, item) => {
    const currentSize = itemSize(item);
    if (currentSize < smallestSize) return currentSize;
    return smallestSize;
  }, firstElementSize);

  if (minSize === 0) {
    console.warn('The size of the smallest item in the list is 0. The list will appear empty.');
  }

  return minSize;
};

export const getNumberOfItemsVisibleOnScreen = <T>({
  data,
  listSizeInPx,
  itemSize,
}: {
  data: T[];
  listSizeInPx: number;
  itemSize: number | ((item: T) => number);
}) => {
  if (data.length === 0) {
    return 0;
  }

  const itemSizeToComputeRanges = getMinSizeOfItems({ data, itemSize });

  if (!itemSizeToComputeRanges) {
    return 0;
  }

  if (itemSizeToComputeRanges === 0) {
    console.warn('The size of the smallest item in the list is 0. The list will appear empty.');
    return 0;
  }

  return Math.floor(listSizeInPx / itemSizeToComputeRanges);
};
