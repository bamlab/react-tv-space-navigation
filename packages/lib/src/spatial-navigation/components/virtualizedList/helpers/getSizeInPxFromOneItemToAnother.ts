/**
 * This function is used to compute the size in pixels of a range of items in a list.
 * If you want the size taken by items from index 0 to 5, you can call this function with
 * start = 0 and end = 5. The size is computed by summing the size of each item in the range.
 * Similarly, if you want to calculate from the other direct, you can call this function
 * with start = 5 and end = 0.
 * @param data The list of items
 * @param itemSizeInPx The size of an item in pixels. It can be a number or a function that takes an item and returns a number.
 * @param start The start index of the range
 * @param end The end index of the range
 * @returns The size in pixels of the range of items
 **/
export const getSizeInPxFromOneItemToAnother = <T>(
  data: T[],
  itemSizeInPx: number | ((item: T) => number),
  start: number,
  end: number,
): number => {
  const startIndex = start < end ? start : end;
  const endIndex = end > start ? end : start;

  if (typeof itemSizeInPx === 'function') {
    return data.slice(startIndex, endIndex).reduce((acc, item) => acc + itemSizeInPx(item), 0);
  }
  return data.slice(startIndex, endIndex).length * itemSizeInPx;
};
