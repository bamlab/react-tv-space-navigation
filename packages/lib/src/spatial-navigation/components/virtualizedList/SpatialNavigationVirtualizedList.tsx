import { useMemo } from 'react';
import { SpatialNavigationNode } from '../Node';
import {
  SpatialNavigationVirtualizedListWithScroll,
  SpatialNavigationVirtualizedListWithScrollProps,
} from './SpatialNavigationVirtualizedListWithScroll';
import { typedMemo } from '../../helpers/TypedMemo';
import { addIndex } from './helpers/addIndex';
import { ItemWithIndex } from './VirtualizedList';

/**
 * Use this component to render horizontally or vertically virtualized lists with spatial navigation
 * This component wraps the virtualized list inside a parent navigation node.
 * */
export const SpatialNavigationVirtualizedList = typedMemo(
  <T,>(props: SpatialNavigationVirtualizedListWithScrollProps<T>) => {
    const indexedData = useMemo(() => addIndex(props.data), [props.data]);

    return (
      <SpatialNavigationNode
        alignInGrid={props.isGrid ?? false}
        orientation={props.orientation ?? 'horizontal'}
      >
        <SpatialNavigationVirtualizedListWithScroll<T & ItemWithIndex>
          {...props}
          data={indexedData}
        />
      </SpatialNavigationNode>
    );
  },
);
SpatialNavigationVirtualizedList.displayName = 'SpatialNavigationVirtualizedList';
