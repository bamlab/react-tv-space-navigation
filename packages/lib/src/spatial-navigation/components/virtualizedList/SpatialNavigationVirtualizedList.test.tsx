import { RenderResult, act, render, screen } from '@testing-library/react-native';
import { ItemWithIndex } from '../virtualizedList/VirtualizedList';
import { PropsTestButton, TestButton } from '../tests/TestButton';
import { SpatialNavigationRoot } from '../Root';
import '../tests/helpers/configureTestRemoteControl';
import { SpatialNavigationVirtualizedList } from './SpatialNavigationVirtualizedList';
import { DefaultFocus } from '../../context/DefaultFocusContext';
import testRemoteControlManager from '../tests/helpers/testRemoteControlManager';
import { setComponentLayoutSize } from '../../../testing/setComponentLayoutSize';

export const expectButtonToHaveFocus = (component: RenderResult, text: string) => {
  const element = component.getByRole('button', { name: text });
  expect(element).toBeSelected();
};

describe('SpatialNavigationVirtualizedList', () => {
  const renderItem = ({ item }: { item: PropsTestButton & ItemWithIndex }) => (
    <TestButton title={item.title} onSelect={item.onSelect} />
  );

  const data = [
    { title: 'button 1', onSelect: () => undefined, index: 0 },
    { title: 'button 2', onSelect: () => undefined, index: 1 },
    { title: 'button 3', onSelect: () => undefined, index: 2 },
    { title: 'button 4', onSelect: () => undefined, index: 3 },
    { title: 'button 5', onSelect: () => undefined, index: 4 },
    { title: 'button 6', onSelect: () => undefined, index: 5 },
    { title: 'button 7', onSelect: () => undefined, index: 6 },
    { title: 'button 8', onSelect: () => undefined, index: 7 },
    { title: 'button 9', onSelect: () => undefined, index: 8 },
    { title: 'button 10', onSelect: () => undefined, index: 9 },
  ];

  const dataWithVariableSizes = [
    { title: 'button 1', onSelect: () => undefined, index: 0 },
    { title: 'button 2', onSelect: () => undefined, index: 1 },
    { title: 'button 3', onSelect: () => undefined, index: 2 },
    { title: 'button 4', onSelect: () => undefined, index: 3 },
    { title: 'button 5', onSelect: () => undefined, index: 4 },
    { title: 'button 6', onSelect: () => undefined, index: 5 },
    { title: 'button 7', onSelect: () => undefined, index: 6 },
    { title: 'button 8', onSelect: () => undefined, index: 7 },
    { title: 'button 9', onSelect: () => undefined, index: 8 },
    { title: 'button 10', onSelect: () => undefined, index: 9 },
  ];

  const itemSize = (item: { title: string; onSelect: () => undefined; index: number }) =>
    item.index % 2 === 0 ? 100 : 200;

  const listTestId = 'test-list';

  const renderList = () =>
    render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedList
            testID={listTestId}
            renderItem={renderItem}
            data={data}
            itemSize={100}
            numberOfRenderedItems={5}
            numberOfItemsVisibleOnScreen={3}
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );

  it('renders the correct number of item', async () => {
    const component = renderList();
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ width: 1000 });

    expect(screen).toMatchSnapshot();

    const button1 = screen.getByText('button 1');
    expect(button1).toBeTruthy();
    expectButtonToHaveFocus(component, 'button 1');

    const button2 = screen.getByText('button 2');
    expect(button2).toBeTruthy();

    const button3 = screen.getByText('button 3');
    expect(button3).toBeTruthy();

    const button4 = screen.getByText('button 4');
    expect(button4).toBeTruthy();

    const button5 = screen.getByText('button 5');
    expect(button5).toBeTruthy();

    const button6 = screen.queryByText('button 6');
    expect(button6).toBeFalsy();
  });

  it('handles correctly RIGHT and RENDERS new elements accordingly while deleting elements that are too far from scroll', async () => {
    const component = renderList();
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });
    expect(listElement).toHaveStyle({ width: 1000 });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -100 }] });

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.queryByText('button 6')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 3');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -200 }] });

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.queryByText('button 7')).toBeFalsy();

    expect(screen).toMatchSnapshot();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 4');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.queryByText('button 8')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 5');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -400 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 6');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -500 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 7');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -600 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 8');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -700 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 9');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -700 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 10');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -700 }] });
  });

  it('handles correctly RIGHT and RENDERS new elements accordingly while deleting elements that are too far from scroll when on stick to end scroll', async () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedList
            testID="test-list"
            renderItem={renderItem}
            data={data}
            itemSize={100}
            numberOfRenderedItems={5}
            numberOfItemsVisibleOnScreen={3}
            scrollBehavior="stick-to-end"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });
    expect(listElement).toHaveStyle({ width: 1000 });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.queryByText('button 6')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 3');
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.queryByText('button 7')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 4');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -100 }] });

    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.queryByText('button 8')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 5');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -200 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 6');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 7');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -400 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 8');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -500 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 9');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -600 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 10');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -700 }] });
  });

  it('handles correctly RIGHT and RENDERS new elements accordingly while deleting elements that are too far from scroll when on jump on scroll', async () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedList
            testID="test-list"
            renderItem={renderItem}
            data={data}
            itemSize={100}
            numberOfRenderedItems={5}
            numberOfItemsVisibleOnScreen={3}
            scrollBehavior="jump-on-scroll"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });
    expect(listElement).toHaveStyle({ width: 1000 });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.queryByText('button 6')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 3');
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.queryByText('button 7')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 4');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.queryByText('button 8')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 5');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 6');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 7');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -600 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 8');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -600 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 9');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -600 }] });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 10');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -700 }] });
  });

  it('handles correctly different item sizes', async () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedList
            testID="test-list"
            renderItem={renderItem}
            data={dataWithVariableSizes}
            itemSize={itemSize}
            numberOfRenderedItems={5}
            numberOfItemsVisibleOnScreen={3}
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });
    expect(listElement).toHaveStyle({ width: 1500 });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -100 }] });

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.queryByText('button 6')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 3');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.queryByText('button 7')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 4');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -400 }] });

    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.queryByText('button 8')).toBeFalsy();
  });

  it('handles correctly different item sizes on stick to end scroll', async () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedList
            testID="test-list"
            renderItem={renderItem}
            data={dataWithVariableSizes}
            itemSize={itemSize}
            numberOfRenderedItems={5}
            numberOfItemsVisibleOnScreen={3}
            scrollBehavior="stick-to-end"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    act(() => jest.runAllTimers());

    setComponentLayoutSize(listTestId, component, { width: 300, height: 300 });

    const listElement = await component.findByTestId(listTestId);
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });
    expect(listElement).toHaveStyle({ width: 1500 });

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expect(listElement).toHaveStyle({ transform: [{ translateX: 0 }] });

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.queryByText('button 6')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 3');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -100 }] });

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.queryByText('button 7')).toBeFalsy();

    await testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 4');
    expect(listElement).toHaveStyle({ transform: [{ translateX: -300 }] });

    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.queryByText('button 8')).toBeFalsy();
  });
});
