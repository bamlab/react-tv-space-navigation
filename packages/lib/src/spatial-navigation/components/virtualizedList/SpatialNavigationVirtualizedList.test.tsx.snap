// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpatialNavigationVirtualizedList handles correctly RIGHT and RENDERS new elements accordingly while deleting elements that are too far from scroll 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "height": "100%",
      "width": "100%",
    }
  }
  testID="test-list-size-giver"
>
  <View
    collapsable={false}
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "transform": [
          {
            "translateX": -200,
          },
        ],
        "width": 1000,
      }
    }
    testID="test-list"
  >
    <View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 1"
          accessibilityLabel="button 1"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 1
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 100,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 2"
          accessibilityLabel="button 2"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 2
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 200,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 3"
          accessibilityLabel="button 3"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": true,
            }
          }
          accessible={true}
          isFocused={true}
          style={
            [
              {
                "backgroundColor": "red",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 3
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 300,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 4"
          accessibilityLabel="button 4"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 4
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 400,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 5"
          accessibilityLabel="button 5"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 5
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 500,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 6"
          accessibilityLabel="button 6"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 6
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 600,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 7"
          accessibilityLabel="button 7"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 7
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SpatialNavigationVirtualizedList renders the correct number of item 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "height": "100%",
      "width": "100%",
    }
  }
  testID="test-list-size-giver"
>
  <View
    collapsable={false}
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 1000,
      }
    }
    testID="test-list"
  >
    <View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 1"
          accessibilityLabel="button 1"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": true,
            }
          }
          accessible={true}
          isFocused={true}
          style={
            [
              {
                "backgroundColor": "red",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 1
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 100,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 2"
          accessibilityLabel="button 2"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 2
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 200,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 3"
          accessibilityLabel="button 3"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 3
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 300,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 4"
          accessibilityLabel="button 4"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 4
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 400,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 5"
          accessibilityLabel="button 5"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 5
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 500,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 6"
          accessibilityLabel="button 6"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 6
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": 600,
              },
            ],
          }
        }
      >
        <View
          accessibilityHint="button 7"
          accessibilityLabel="button 7"
          accessibilityRole="button"
          accessibilityState={
            {
              "selected": false,
            }
          }
          accessible={true}
          isFocused={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderRadius": 100,
                "padding": 6,
              },
              undefined,
            ]
          }
        >
          <Text>
            button 7
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
